src/test.cpp

```diff
@@ -0,0 +1,297 @@
+                // Exception for bits==0
+                if (bits == 0) CHECK(table_capacity[max_elements] == 0);
+                // To be able to decode N elements, capacity needs to be at least N.
+                if (bits > 0) CHECK(table_capacity[max_elements] >= max_elements);
+                // A sketch of N bits in total cannot have more than N bits of false positive protection;
+                if (bits > 0) CHECK(bits * table_capacity[max_elements] >= fpbits);
+                // When asking for N bits of false positive protection, no more than ceil(N / bits) excess capacity should be needed.
+                if (bits > 0) CHECK(table_capacity[max_elements] - max_elements <= (fpbits + bits - 1) / bits);
+                // Increasing max_elements by one can only increment the capacity by 0 or 1.
+                if (max_elements > 0 && fpbits < 256) CHECK(table_capacity[max_elements] == table_capacity[max_elements - 1] || table_capacity[max_elements] == table_capacity[max_elements - 1] + 1);
+                // Check round-tripping max_elements->capacity->max_elements (only a lower bound)
+                CHECK(table_capacity[max_elements] <= 1024);
+                CHECK(table_max_elements[table_capacity[max_elements]] == 0 || table_max_elements[table_capacity[max_elements]] >= max_elements);
+            }
+
+            for (size_t capacity = 0; capacity <= 512; ++capacity) {
+                // Check round-tripping capacity->max_elements->capacity (exact, if it exists)
+                CHECK(table_max_elements[capacity] <= 512);
+                CHECK(table_max_elements[capacity] == 0 || table_capacity[table_max_elements[capacity]] == capacity);
+            }
+        }
+    }
+}
+
+} // namespace
+
+int main(int argc, char** argv) {
+    uint64_t test_complexity = 4;
+    if (argc > 1) test_complexity = std::stoull(std::string(argv[1]), nullptr);
+    if (test_complexity <= 0) {
+        fprintf(stderr, "Iteration count must be > 0.");
+        return 1;
+    }
+#ifdef MINISKETCH_VERIFY
+    const char* mode = " in verify mode";
+#else
+    const char* mode = "";
+#endif
+    printf("Running libminisketch tests%s with complexity=%llu\n", mode, (unsigned long long)test_complexity);
+
+    TestComputeFunctions();
+
+    for (unsigned j = 2; j <= 64; j += 1) {
+        TestRandomized(j, 8, (test_complexity << 10) / j);
+        TestRandomized(j, 128, (test_complexity << 7) / j);
+        TestRandomized(j, 4096, test_complexity / j);
+    }
+
+    for (int weight = 2; weight <= 40; weight += 1) {
+        for (int bits = 2; bits <= 32 && bits <= weight; ++bits) {
+            int capacity = weight / bits;
+            if (capacity * bits != weight) continue;
+            TestExhaustive(bits, capacity);
+        }
+        if (weight >= 16 && test_complexity >> (weight - 16) == 0) break;
+    }
+
+    printf("All tests succesful.\n");
```

Comment:
s/succesful/successful/

Reply:
Done.

Reply:
Seems I duplicated the wrong letter here.